File: src/main/java/com/github/takemikami/intellij/plugin/externallinters/python/Flake8LocalInspection.java
package com.github.takemikami.intellij.plugin.externallinters.python;

import com.github.takemikami.intellij.plugin.externallinters.LinterProblem;
import com.intellij.codeInspection.ProblemsHolder;
import com.intellij.psi.PsiElementVisitor;
import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.jetbrains.annotations.NotNull;

/**
 * Local Inspection of flake8.
 */
public class Flake8LocalInspection extends AbstractPythonInspection {

  @NotNull
  @Override
  public PsiElementVisitor buildVisitor(@NotNull ProblemsHolder holder, boolean isOnTheFly) {
    return buildPsiElementVisitorByCommand(holder, isOnTheFly, "pflake8");
  }

  boolean isEnabledByPyprojectToml(String tomlBody) {
    return Arrays.stream(tomlBody.split("\n"))
        .anyMatch(ln -> ln.trim().startsWith("[tool.flake8]"));
  }

  private static final Pattern OUTPUT_PATTERN = Pattern.compile(
      "([^:]*):([^:]*):([^:]*):\\s*([A-Z0-9]*)\\s*(.*)");

  List<LinterProblem> run(String binPath, String basePath, String path, String body) throws IOException {
    return runLinter(
        binPath,
        new String[]{
            "--format",
            "'%(path)s:%(row)d:%(col)d: %(code)s %(text)s'",
            "-"
        },
        basePath,
        null,
        path,
        body,
        OUTPUT_PATTERN
    );
  }

  LinterProblem createLinterProblemByMatcher(@NotNull Matcher matcher) {
    return new LinterProblem(
        matcher.group(1),
        Integer.parseInt(matcher.group(2)),
        Integer.parseInt(matcher.group(3)),
        matcher.group(4),
        null,
        matcher.group(5)
    );
  }

}

File: src/main/kotlin/com/github/takemikami/intellij/plugin/pyprojectlinters/Flake8LocalInspection.kt
package com.github.takemikami.intellij.plugin.pyprojectlinters

import com.intellij.codeInspection.ProblemsHolder
import com.intellij.psi.PsiElementVisitor
import java.util.regex.Matcher
import java.util.regex.Pattern

/**
 * Local Inspection of flake8.
 */
class Flake8LocalInspection : AbstractPythonInspection() {
    override fun buildVisitor(
        holder: ProblemsHolder,
        isOnTheFly: Boolean,
    ): PsiElementVisitor {
        return buildPsiElementVisitorByCommand(holder, isOnTheFly, "pflake8")
    }

    override fun isEnabledByPyprojectToml(tomlBody: String): Boolean {
        return tomlBody.split("\n").stream()
            .anyMatch { ln -> ln.trim().startsWith("[tool.flake8]") }
    }

    val outputPattern: Pattern =
        Pattern.compile(
            "([^:]*):([^:]*):([^:]*):\\s*([A-Z0-9]*)\\s*(.*)",
        )

    override fun run(
        binPath: String,
        basePath: String,
        path: String,
        body: String,
    ): List<LinterProblem> {
        return runLinter(
            binPath,
            arrayOf<String>(
                "--format",
                "'%(path)s:%(row)d:%(col)d: %(code)s %(text)s'",
                "-",
            ),
            basePath,
            null,
            path,
            body,
            outputPattern,
        )
    }

    override fun createLinterProblemByMatcher(matcher: Matcher): LinterProblem? {
        val msg = matcher.group(4) + " " + matcher.group(5)
        return LinterProblem(
            matcher.group(1),
            matcher.group(2).toInt(),
            matcher.group(3).toInt(),
            msg,
        )
    }
}
